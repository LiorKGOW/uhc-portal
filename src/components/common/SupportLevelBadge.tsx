import React from 'react';
import { Popover, PopoverPosition, Label } from '@patternfly/react-core';
import { InfoCircleIcon } from '@patternfly/react-icons';

import links from '../../common/installLinks.mjs';
import ExternalLink from './ExternalLink';

export enum SupportLevelType {
  devPreview = 'DEV_PREVIEW',
  cooperativeCommunity = 'COOPERATIVE_COMMUNITY',
}

type BadgeSupportData = {
  [key in SupportLevelType]: {
    title: string;
    link: string;
    text: string;
  };
};

const badgeLevels: BadgeSupportData = {
  [SupportLevelType.devPreview]: {
    title: 'Developer Preview',
    link: links.INSTALL_PRE_RELEASE_SUPPORT_KCS,
    text:
      'Developer preview features provide early access to upcoming product innovations, enabling\n' +
      'you to test functionality and provide feedback during the development process.',
  },
  [SupportLevelType.cooperativeCommunity]: {
    title: 'Cooperative Community',
    link: links.COOPERATIVE_COMMUNITY_SUPPORT_KCS,
    text:
      'Cooperative Community Support provides assistance to Red Hat customers that have questions\n' +
      'about community-provided software that is often used with our Red Hat products',
  },
};

const SupportLevelBadge = ({ type }: { type: SupportLevelType }) => {
  const badgeInfo = badgeLevels[type];
  const infoElem = (
    <>
      <div className="pf-u-mb-sm">{badgeInfo.text}</div>
      <ExternalLink href={badgeInfo.link}>Learn more</ExternalLink>
    </>
  );
  return (
    <Popover position={PopoverPosition.top} bodyContent={infoElem}>
      <Label
        style={{ cursor: 'pointer' }}
        color="orange"
        onClick={(event) => {
          event.preventDefault();
        }}
        icon={<InfoCircleIcon color="var(--pf-c-label__content--Color)" />}
        className="pf-u-ml-md pf-u-display-inline"
      >
        {badgeInfo.title}
      </Label>
    </Popover>
  );
};

export default SupportLevelBadge;
