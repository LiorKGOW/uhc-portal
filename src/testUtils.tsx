import React from 'react';
import { Provider } from 'react-redux';
import { createStore, Store } from 'redux';
import { render, RenderOptions } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import { toHaveNoViolations, axe } from 'jest-axe';

import { createBrowserHistory } from 'history';
import { store } from './redux/store';
import { reduxReducers } from './redux/reducers';
import * as restrictedEnv from './restrictedEnv';

const history = createBrowserHistory();

export const TestWrapper = ({
  children,
  initialStore,
}: {
  children: React.ReactNode;
  initialStore?: Store;
}) => <Provider store={initialStore || store}>{children}</Provider>;

const renderWithState = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
  initialState?: any, // setting to any because only a partial state structure may be sent
) => {
  const newStore = initialState ? createStore(reduxReducers(history), initialState) : store;

  const Wrapper = ({ children }: { children: React.ReactNode }) => (
    <TestWrapper initialStore={newStore}>{children}</TestWrapper>
  );
  return { ...render(ui, { wrapper: Wrapper, ...options }), user: userEvent.setup() };
};

export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';

// Type not exported in the library
export type UserEventType = {
  type: (
    field: HTMLElement,
    typeValue: string,
    { initialSelectionStart }: { initialSelectionStart: number },
  ) => void;
  clear: (field: HTMLElement) => void;
};

export { renderWithState as render };

/* ***** Items outside of React Test Library ************ */
expect.extend(toHaveNoViolations);

export const checkAccessibility = async (container: HTMLElement | string) => {
  const results = await axe(container);
  expect(results).toHaveNoViolations();
};

export const insightsMock = () => {
  global.insights = {
    chrome: {
      ...global.insights.chrome,
      on: () => () => {},
      appNavClick: () => {},
      auth: {
        getUser: () => Promise.resolve({ data: {} }),
        getToken: () => Promise.resolve(),
        getOfflineToken: () => Promise.resolve({ data: { refresh_token: 'hello' } }),
      },
    },
  };
};

export const mockRestrictedEnv = () => {
  const mock = jest.spyOn(restrictedEnv, 'isRestrictedEnv');
  mock.mockReturnValue(false);
  return mock;
};
