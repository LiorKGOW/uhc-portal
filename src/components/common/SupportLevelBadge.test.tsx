import React from 'react';

import { render, screen, within } from '~/testUtils';

import SupportLevelBadge, { COOPERATIVE_COMMUNITY, DEV_PREVIEW } from './SupportLevelBadge';

describe('<SupportLevelBadge>', () => {
  it('displays "Developer Preview" for dev preview badges', async () => {
    const { user } = render(<SupportLevelBadge {...DEV_PREVIEW} />);
    expect(screen.getByText('Developer Preview')).toBeInTheDocument();

    await user.click(screen.getByRole('button'));

    expect(await screen.findByRole('dialog')).toBeInTheDocument();

    expect(within(screen.getByRole('dialog')).getByRole('link')).toHaveAttribute(
      'href',
      'https://access.redhat.com/support/offerings/devpreview',
    );
    expect(
      within(screen.getByRole('dialog')).getByText(
        'Developer preview features provide early access to upcoming product innovations',
        { exact: false },
      ),
    ).toBeInTheDocument();
  });

  it('displays "Cooperative Community" for cooperative community badges', async () => {
    const { user } = render(<SupportLevelBadge {...COOPERATIVE_COMMUNITY} />);
    expect(screen.getByText('Cooperative Community')).toBeInTheDocument();

    await user.click(screen.getByRole('button'));

    expect(await screen.findByRole('dialog')).toBeInTheDocument();

    expect(within(screen.getByRole('dialog')).getByRole('link')).toHaveAttribute(
      'href',
      'https://access.redhat.com/solutions/5893251',
    );
    expect(
      within(screen.getByRole('dialog')).getByText(
        'Cooperative Community Support provides assistance to Red Hat customers',
        { exact: false },
      ),
    ).toBeInTheDocument();
  });

  it('without externalLink omits "Learn more"', async () => {
    const content = (
      <>
        Foo <b>bar</b> baz
      </>
    );
    const { user } = render(<SupportLevelBadge text="text" popoverContent={content} />);

    await user.click(screen.getByRole('button'));

    expect(await screen.findByRole('dialog')).toBeInTheDocument();
    expect(within(screen.getByRole('dialog')).getByText('bar')).toBeInTheDocument();
    expect(within(screen.getByRole('dialog')).queryByText('Learn more')).not.toBeInTheDocument();
  });
});
