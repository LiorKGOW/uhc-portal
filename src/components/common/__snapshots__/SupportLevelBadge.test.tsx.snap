// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SupportLevelBadge> renders for cooperative community badges 1`] = `
<Popover
  bodyContent={
    <React.Fragment>
      <div
        className="pf-v5-u-mb-sm"
      >
        Cooperative Community Support provides assistance to Red Hat customers that have questions
about community-provided software that is often used with our Red Hat products
      </div>
      <ExternalLink
        href="https://access.redhat.com/solutions/5893251"
      >
        Learn more
      </ExternalLink>
    </React.Fragment>
  }
  position="top"
>
  <Label
    className="pf-v5-u-ml-md pf-v5-u-display-inline"
    color="orange"
    icon={
      <Icon>
        <InfoCircleIcon
          color="var(--pf-v5-c-label__content--Color)"
        />
      </Icon>
    }
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
  >
    Cooperative Community
  </Label>
</Popover>
`;

exports[`<SupportLevelBadge> renders for devPreview badges 1`] = `
<Popover
  bodyContent={
    <React.Fragment>
      <div
        className="pf-v5-u-mb-sm"
      >
        Developer preview features provide early access to upcoming product innovations, enabling
you to test functionality and provide feedback during the development process.
      </div>
      <ExternalLink
        href="https://access.redhat.com/support/offerings/devpreview"
      >
        Learn more
      </ExternalLink>
    </React.Fragment>
  }
  position="top"
>
  <Label
    className="pf-v5-u-ml-md pf-v5-u-display-inline"
    color="orange"
    icon={
      <Icon>
        <InfoCircleIcon
          color="var(--pf-v5-c-label__content--Color)"
        />
      </Icon>
    }
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
  >
    Developer Preview
  </Label>
</Popover>
`;
